query/filter/key_builder.go

```diff
@@ -121,56 +226,176 @@ func (s *StrictEqKeyComposer) Compose(selectors []*Selector, userDefinedKeys []*
-			keyPartsCopy := make([]*Selector, len(compositeKeys[0])-1)
-			copy(keyPartsCopy, compositeKeys[0][0:len(compositeKeys[0])-1])
-			keyPartsCopy = append(keyPartsCopy, repeatedFields[j]) //nolint:makezero
-			compositeKeys = append(compositeKeys, keyPartsCopy)    //nolint:makezero
+		if s.matchAll {
+			compositeKeys[0] = append(compositeKeys[0], repeatedFields[0])
+			// as we found some repeated fields in the filter so clone the first set of keys and add this prefix to all the
+			// repeated fields, cloning is only needed if there are more than one repeated fields
+			for j := 1; j < len(repeatedFields); j++ {
+				keyPartsCopy := make([]*Selector, len(compositeKeys[0])-1)
+				copy(keyPartsCopy, compositeKeys[0][0:len(compositeKeys[0])-1])
+				keyPartsCopy = append(keyPartsCopy, repeatedFields[j]) //nolint:makezero
+				compositeKeys = append(compositeKeys, keyPartsCopy)    //nolint:makezero
+			}
+		} else {
+			compositeKeys = append(compositeKeys, [][]*Selector{repeatedFields}...) //nolint:makezero
 		}
 	}
 
 	// keys building is dependent on the filter type
-	var allKeys []keys.Key
+	var queryPlans []QueryPlan
 	for _, k := range compositeKeys {
 		switch parent {
 		case AndOP:
-			var primaryKeyParts []interface{}
-			for _, s := range k {
-				primaryKeyParts = append(primaryKeyParts, s.Matcher.GetValue().AsInterface())
+			var keyParts []interface{}
+			for _, sel := range k {
+				newParts := s.buildIndexPartsFunc(sel.Field.Name(), sel.Field.DataType, sel.Matcher.GetValue().AsInterface())
+				keyParts = append(keyParts, newParts...)
 			}
 
-			key, err := s.keyEncodingFunc(primaryKeyParts...)
+			key, err := s.keyEncodingFunc(keyParts...)
 			if err != nil {
 				return nil, err
 			}
-			allKeys = append(allKeys, key)
+			dataType := schema.UnknownType
+			if len(k) == 1 {
+				dataType = k[0].Field.DataType
+			}
+			queryPlans = append(queryPlans, newQueryPlan(EQUAL, dataType, []keys.Key{key}))
 		case OrOP:
 			for _, sel := range k {
 				if len(userDefinedKeys) > 1 {
 					// this means OR can't build independently these keys
 					return nil, errors.InvalidArgument("OR is not supported with composite primary keys")
 				}
 
-				key, err := s.keyEncodingFunc(sel.Matcher.GetValue().AsInterface())
+				primaryKeyParts := s.buildIndexPartsFunc(sel.Field.Name(), sel.Field.DataType, sel.Matcher.GetValue().AsInterface())
+
+				key, err := s.keyEncodingFunc(primaryKeyParts...)
 				if err != nil {
 					return nil, err
 				}
 
-				allKeys = append(allKeys, key)
+				queryPlans = append(queryPlans, newQueryPlan(EQUAL, sel.Field.DataType, []keys.Key{key}))
 			}
 		}
 	}
 
-	return allKeys, nil
+	return queryPlans, nil
+}
+
+// Range Key Composer will generate a range key set on the user defined keys
+// It will set the KeyQuery to `FullRange` if the start or end key is not defined in the query
+// if there is a defined start and end key for a range then `Range` is set.
+type RangeKeyComposer[F fieldable] struct {
```

Comment:
For this diff, the structure is good. But if we think about it, we just need two high level abstractions here,
  - `NewQueryableKeyBuilder(composer, primaryKey)` [will be used by insert/replace only]
  - `NewQueryableKeyBuilder()`: This will internally use eq/range composer to build plans i.e. equality as well as range. The reason being once we don't rely on the search store then either the plan is formed or we need to fall back to full table scan, there will be no other option. Then `Read/Update/Delete` can rely directly on it. This means we need to add something to the QueryPlan so that caller knows whether the keys are for primary or secondary indexes.

Reply:
Yes I like it. We can definitely move to that direction once search and secondary are ready.

Reply:
sg